From 27ae87a59cb2f0d06ea21d4a886f3afb9a14247d Mon Sep 17 00:00:00 2001
From: "Gary D. Gregory" <ggregory@apache.org>
Date: Tue, 9 Oct 2018 18:44:18 +0000
Subject: [PATCH] [CONFIGURATION-727]
 org.apache.commons.configuration2.DatabaseConfiguration never closes result
 sets and statements.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1843324 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                       |  2 +-
 .../configuration2/DatabaseConfiguration.java | 26 +++++++++++--------
 2 files changed, 16 insertions(+), 12 deletions(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index dd602afcaa..4d741b662f 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -79,7 +79,7 @@
         Add support for Commons Text 1.5 new string lookups as default lookups.
       </action>
       <action dev="ggregory" type="fix" issue="CONFIGURATION-727">
-        org.apache.commons.configuration2.DatabaseConfiguration never closes result sets.
+        org.apache.commons.configuration2.DatabaseConfiguration never closes result sets and statements.
       </action>
     </release>
 
diff --git a/src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java b/src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java
index 193b1f407c..5165737dd8 100644
--- a/src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java
@@ -390,15 +390,17 @@ protected Void performOperation() throws SQLException
                 }
                 query.append(")");
 
-                final PreparedStatement pstmt = initStatement(query.toString(),
-                        false, key, String.valueOf(obj));
-                if (configurationNameColumn != null)
+                try (final PreparedStatement pstmt = initStatement(query.toString(),
+                        false, key, String.valueOf(obj)))
                 {
-                    pstmt.setString(3, configurationName);
-                }
+                    if (configurationNameColumn != null)
+                    {
+                        pstmt.setString(3, configurationName);
+                    }
 
-                pstmt.executeUpdate();
-                return null;
+                    pstmt.executeUpdate();
+                    return null;
+                }
             }
         }
         .execute();
@@ -512,10 +514,12 @@ protected void clearPropertyDirect(final String key)
             @Override
             protected Void performOperation() throws SQLException
             {
-                final PreparedStatement ps = initStatement(String.format(
-                        SQL_CLEAR_PROPERTY, table, keyColumn), true, key);
-                ps.executeUpdate();
-                return null;
+                try (final PreparedStatement ps = initStatement(String.format(
+                        SQL_CLEAR_PROPERTY, table, keyColumn), true, key))
+                {
+                    ps.executeUpdate();
+                    return null;
+                }
             }
         }
         .execute();
